144 loadnifti transform common dataset

Nic-Ma opened the pull request:
Fixes #144 #146 .

### Description
This PR implemented:
1. Vanilla level and dictionary level transforms of LoadNifti.
2. Common Dataset for dictionary format input data.
3. Updated 3D segmentation example to the latest dictionary-based APIs.
4. Added ChangeToChannelFirst transform.
5. Fixed bug in Dice loss.

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

wyli commented on monai/transforms/transforms.py:
```diff
@@ -24,6 +25,65 @@
 
 
+@export
+class LoadNifti:
```
could we also have an additional `LoadNiftiMultichannel` so that we can stack multiple files at the first dim (channel dim)?

Nic-Ma replied:
Hi @wyli ,

I think that should be another transform: "MergeToChannels", we have it in Clara.
Thanks.

wyli commented on monai/data/nifti_reader.py:
```diff
@@ -135,3 +135,38 @@ def __getitem__(self, index):
+    def __init__(self, data, transform=None):
+        """
+        Args:
+            data (dict): input data to load and transform to generate dataset for model.
```
typo: data (list of dictionaries)

Nic-Ma replied:
updated in latest patch.

wyli commented on monai/transforms/transforms.py:
```diff
@@ -24,6 +25,65 @@
+        if self.dtype is not None:
+            data = data.get_fdata(dtype=self.dtype)
+        else:
+            data = np.asanyarray(data.dataobj)
```
now it's a proper transform, we need to make the data in the channel-first format

Nic-Ma replied:
As the offline discussion, will develop another transform: ChangeToChannelFirst.
Thanks.

wyli commented on monai/transforms/transforms.py:
```diff
@@ -24,6 +25,65 @@
+        self.image_only = image_only
+        self.dtype = dtype
+
+    def __call__(self, img):
```
variable name could be image_name or filename

Nic-Ma replied:
Updated in the latest patch.
Thanks.

wyli commented on monai/data/nifti_reader.py:
```diff
@@ -135,3 +135,38 @@ def __getitem__(self, index):
+         'img': 'image1.nii.gz',      'img': 'image2.nii.gz',      'img': 'image3.nii.gz',
+         'seg': 'label1.nii.gz',      'seg': 'label2.nii.gz',      'seg': 'label3.nii.gz',
+         'extra': 123,                'extra': 456,                'extra': 789,
+         'shape': 'CHWD'              'shape': 'CHWD'              'shape': 'CHWD'
```
the "shape' in the doc can be misleading, could you remove?

Nic-Ma replied:
Updated in the latest patch.
Thanks.

wyli commented on monai/data/nifti_reader.py:
```diff
@@ -135,3 +135,38 @@ def __getitem__(self, index):
+            data (dict): input data to load and transform to generate dataset for model.
+            transform (Callable, optional): dict transforms to excute operations on dictionary data.
+        """
+        assert isinstance(data, list) and all(isinstance(item, dict) for item in data), \
```
I think the pythonic way here would be not restricting the types, as long as `data[index]` returns some data, there's no need to make type assumption. For example `data` could even be a `torch.utils.data.dataset` object, as long as `self.data[index]` returns something we don't raise an error at this level.

Nic-Ma replied:
Updated in the latest patch.
Thanks.

Nic-Ma commented:
Hi @wyli @ericspod @atbenmurray ,

I updated code according to the comments.
Could you please help review it again?
Thanks.

madil90 commented on examples/unet_segmentation_3d_dict.py:
```diff
@@ -52,15 +52,22 @@
 images = sorted(glob(os.path.join(tempdir, 'im*.nii.gz')))
 segs = sorted(glob(os.path.join(tempdir, 'seg*.nii.gz')))
+train_data = list()
+for img, seg in zip(images[:30], segs[:30]):
```
can be rewritten as 
```suggestion
train_data = [{'img': img, 'seg': seg} for img, seg in zip(images[:30], segs[:30])]
```
more pythonic.

Nic-Ma replied:
Updated in the latest patch.
Thanks.

madil90 commented on examples/unet_segmentation_3d_dict.py:
```diff
@@ -52,15 +52,22 @@
+for img, seg in zip(images[:30], segs[:30]):
+    train_data.append({'image': img, 'seg': seg})
+val_data = list()
+for img, seg in zip(images[-20:], segs[-20:]):
```
Similarly
```suggestion
train_data = [{'img': img, 'seg': seg} for img, seg in zip(images[-20:], segs[-20:])]
```

Nic-Ma replied:
Updated in the latest patch.
Thanks.

madil90 commented on monai/data/dataset.py:
```diff
@@ -0,0 +1,45 @@
+    General Dataset to handle dictionary format data, it can operate transforms for specific fields.
+    For example, typical input data can be a list of dictionaries:
+    [{                            {                            {
+         'img': 'image1.nii.gz',      'img': 'image2.nii.gz',      'img': 'image3.nii.gz',
```
Does it make sense to ask user to replicate the 'img' key. I expect most users will have data in format {'img': images, 'label': labels}. @wyli @ericspod can you share your thoughts?

madil90 commented on monai/transforms/transforms.py:
```diff
@@ -26,6 +28,65 @@
+            if type(meta_datum).__name__ == 'ndarray' \
+                    and np_str_obj_array_pattern.search(meta_datum.dtype.str) is not None:
+                continue
+            compatible_meta[meta_key] = meta_datum
```
nifti_reader already has this logic. Why replicate it here?

Nic-Ma replied:
Hi @madil90 ,

Load Nifti transform is more flexible, will also discuss with Eric to remove nifti_reader later.
Thanks.

madil90 commented on tests/test_dataset.py:
```diff
@@ -0,0 +1,75 @@
+    @parameterized.expand([TEST_CASE_1])
+    def test_shape(self, input_param, expected_shape):
+        test_image = nib.Nifti1Image(np.random.randint(0, 2, size=[128, 128, 128]), np.eye(4))
+        nib.save(test_image, 'test_image1.nii.gz')
```
Create a tmp dir to save this. Safer than saving in local dir (if program halts etc).

Nic-Ma replied:
Updated in the latest patch.
Thanks.

madil90 commented on tests/test_load_nifti.py:
```diff
@@ -0,0 +1,53 @@
+    @parameterized.expand([TEST_CASE_1, TEST_CASE_2])
+    def test_shape(self, input_param, input_data, expected_shape):
+        test_image = np.random.randint(0, 2, size=[128, 128, 128])
+        nib.save(nib.Nifti1Image(test_image, np.eye(4)), 'test_image.nii.gz')
```
Should be saved in tmp dir.

Nic-Ma replied:
Updated in the latest patch.
Thanks.

madil90 commented on tests/test_load_nifti.py:
```diff
@@ -0,0 +1,53 @@
+    (128, 128, 128)
+]
+
+TEST_CASE_2 = [
```
NAME these TEST_IMAGE_ONLY (or as you like) instead of 1,2. Makes it more readable what this case is for.

Nic-Ma replied:
Updated in the latest patch.
Thanks.

madil90 commented on tests/test_load_niftid.py:
```diff
@@ -0,0 +1,54 @@
+    @parameterized.expand([TEST_CASE_1])
+    def test_shape(self, input_param, input_data, expected_shape):
+        test_image = nib.Nifti1Image(np.random.randint(0, 2, size=[128, 128, 128]), np.eye(4))
+        nib.save(test_image, 'test_image.nii.gz')
```
Save in temp dir.

madil90 commented on tests/test_load_niftid.py:
```diff
@@ -0,0 +1,54 @@
+    @parameterized.expand([TEST_CASE_1])
+    def test_shape(self, input_param, input_data, expected_shape):
+        test_image = nib.Nifti1Image(np.random.randint(0, 2, size=[128, 128, 128]), np.eye(4))
+        nib.save(test_image, 'test_image.nii.gz')
```
Should save to tmp dir.

Nic-Ma replied:
Updated in the latest patch.
Thanks.

Nic-Ma commented on examples/unet_segmentation_3d_dict.py:
```diff
@@ -52,15 +52,22 @@
 segs = sorted(glob(os.path.join(tempdir, 'seg*.nii.gz')))
+train_data = list()
+for img, seg in zip(images[:30], segs[:30]):
+    train_data.append({'image': img, 'seg': seg})
```
Hi @wyli ,

I think we ask the input data to be a dict and no need to be file names.
Users can pass data to the Dataset directly instead of loading from file.
Thanks.

madil90 replied:
I agree with Wenqi. I think user should just be able to pass image file names and label file names. This extra verbosity should be handled by us. Nobody has their data in this Dict form. Why enforce this when we can handle it ourselves.

Nic-Ma replied:
Updated in the latest patch.
Changed to train_files, val_files
Thanks.

Nic-Ma commented on examples/unet_segmentation_3d_dict.py:
```diff
@@ -28,8 +28,8 @@
 import monai.transforms.compose as transforms
-from monai.data.nifti_reader import NiftiDatasetd
-from monai.transforms.composables import AddChanneld, RandRotate90d
+from monai.data.dataset import Dataset
```
Updated in the latest patch.
Thanks.

Nic-Ma commented on monai/transforms/composables.py:
```diff
@@ -53,6 +53,35 @@ def __init__(self, keys):
+            d[key] = data[0]
+            assert isinstance(data[1], dict), 'metadata must be in dict format.'
+            for k, v in data[1].items():
+                d[key + '.' + k] = v
```
Updated in the latest patch.
If self.overwriting_keys is False, it will raise exception directly if matching an existing key.
Thanks.

Nic-Ma commented:
Hi @wyli @madil90 @ericspod @atbenmurray ,

I updated the code according to the comments.
And also added below features to this PR:
1. Added ChangeToChannelFirst transform.
2. Updated 3D segmentation example to the latest dictionary-based APIs.
    Used **ChangeToChannelFirstd, RandCropByPosNegLabeld, RandRotate90d** transforms.
3. Fixed bug in Dice loss: remove smooth from top, otherwise it may generate negative loss value.

I think we need another separate PR to update inference related logic.
Could you please help review it again?
Thanks.

wyli commented on monai/transforms/transforms.py:
```diff
@@ -26,6 +28,84 @@
+    def __init__(self, channel_dim=-1):
+        self.channel_dim = channel_dim
+
+    def __call__(self, img):
```
this could be implemented with `np.moveaxis`

Nic-Ma replied:
Thanks for your suggestion, replaced with np.moveaxis() API.

madil90 commented on examples/unet_segmentation_3d_dict.py:
```diff
@@ -28,42 +28,52 @@
-images = sorted(glob(os.path.join(tempdir, 'im*.nii.gz')))
+images = sorted(glob(os.path.join(tempdir, 'img*.nii.gz')))
 segs = sorted(glob(os.path.join(tempdir, 'seg*.nii.gz')))
+train_files = [{'img': img, 'seg': seg} for img, seg in zip(images[:40], segs[:40])]
```
It's safer to set a variable and use that instead of 'img' and 'seg' everywhere.

Nic-Ma replied:
Wenqi suggested to unify all constant values after GTC.
Including image properties, networks, etc.
Thanks.

madil90 commented on monai/data/synthetic.py:
```diff
@@ -14,7 +14,7 @@
+def create_test_image_2d(width, height, num_objs=12, rad_max=30, noise_max=0.0, num_seg_classes=5, channel_dim=None):
     """
     Return a noisy 2D image with `numObj' circles and a 2D mask image. The maximum radius of the circles is given as
     `radMax'. The mask will have `numSegClasses' number of classes for segmentations labeled sequentially from 1, plus a
```
Add 'channel_dim' explanation to docstring.

Nic-Ma replied:
Updated in the latest patch.
Thanks.

madil90 commented on monai/data/synthetic.py:
```diff
@@ -40,10 +40,16 @@ def create_test_image_2d(width, height, num_objs=12, rad_max=30, noise_max=0.0,
 
+    if channel_dim is not None:
+        assert isinstance(channel_dim, int) and channel_dim in (-1, 0, 2), 'invalid channel dim.'
+        noisyimage, labels = noisyimage[None], labels[None] \
```
Is the point of channel_dim to create channels_first or channels_last image? If yes, then channel_dim int variable is not needed (It's not being used either). Can just have a boolean flag. Also safer for higher dimensional spatial data.

Nic-Ma replied:
It's not a bool, can have 3 types: None, channel first, channel last.
If channel_dim = None, will create an image without channel dimension.
Thanks.

madil90 commented on monai/transforms/transforms.py:
```diff
@@ -26,6 +28,81 @@
+            filename (str or file): path to file or file-like object.
+
+        Returns:
+            The loaded image volume if `image_only` is True, or a tuple containing the volume and the Nifti
```
This should be moved or included in class or init docstring. During usage, when user writes this transform as part of Compose, they will never see this documentation.

Nic-Ma replied:
Updated in the latest patch.
Thanks.

madil90 commented on tests/test_load_niftid.py:
```diff
@@ -0,0 +1,52 @@
+    def test_shape(self, input_param, expected_shape):
+        test_image = nib.Nifti1Image(np.random.randint(0, 2, size=[128, 128, 128]), np.eye(4))
+        tempdir = tempfile.mkdtemp()
+        nib.save(test_image, os.path.join(tempdir, 'test_image.nii.gz'))
```
Can make loops or list comprehensions out of these. Makes our codebase look elegant. Not mandatory. Just a suggestion.

Nic-Ma replied:
Updated in the latest patch.
Thanks.
